### DISCUSSION TITLE
SafeG64：Toppersの更新方法につきまして

### DISCUSSION THREAD

**Discussion by Explorer-kato** [toppers/users-forum#128] [Q&A]
ご担当者様

お世話になっております。
株式会社エクスプローラの加藤と申します。

お忙しいところ、申し訳ございませんがご教示お願いいたします。
[求められている動作]
・SafeG64でLinuxがToppers上で動作するプログラムを更新し
　Linuxを止めることなくToppersのみを再起動し、更新したプログラムで動作させる。
[方針案]
①Linux/Toppers両方でアクセス可能な領域にToppers全体を書き込み
　再起動時に一度Toppersを終了させるときに、その領域にジャンプするようにして
　そこから改めてToppersを動作させる。
②Toppersのカーネルとアプリのデータとプログラムの配置位置を分割し
　更新前にToppersをカーネル領域に退避させる。
　アプリの領域に更新したアプリのデータを更新し、再度アプリのプログラムを実行する。
[仕様パッケージ]
SafeG64：safeg64_1.2.0　Toppers：fmp3_3.1
[質問内容]
・そもそも求められている動作は実現可能でしょうか？
　その場合①、②のどちらが容易でしょうか？
・①の案の場合、現状SafeG64のサンプル動作中にToppers側のアプリを’Q’コマンドで終了させるとLinuxも停止いたしますが、
　更新前のToppers停止処理でLinuxを止めずにマルチプロセッサのToppersが再起動できるようにするための終了処理をご教示いただけないでしょうか？
・②の案の場合、アプリ用のプログラム、データ領域を定義して配置する必要があると思いますが
　セクションの追加方法についてご教示いただけないでしょうか？
　現状imx8mm_evk.ldの編集をしても生成されたelfファイルに追加したセクション情報は無く
　ATT_SECは使用不可という解釈でおります。可能でしたら実例をご教示いただけますと幸いです。

よろしくお願いいたします。

**Comment by exshonda** [comment]
実施したことはありませんが，技術的にはおそらく可能だと思います．
アプリケーションのみ更新されたいと見受けられますので，②が実現できればよいかと思いますが，現状のFMP3にはアプリを更新する機能がありません．

そうすると，①で実現する必要があります．新しいバイナリへのジャンプは，target_exit()に新しくロードしたToppersのエントリへのジャンプ処理を入れれば実現可能です．
その上で，大きく次の3種類の対応が必要になるかと思います．
・ハードウェエアの初期化処理
   ・現状のToppersはコールドブートしたとしてハードウェアを全て初期化します．初期化によってはハングアップしたりLinuxに影響を与えるために注意が必要です．気になるハードウェアについては以下に列挙します．target_initialize()からのコードを精査する必要があります．
      ・キャッシュ初期化（MMUテーブル），割込みコントローラ初期化
・ATFの初期化処理
   btaskは起動するとatf_smc_init()を呼び出しますが，これは必要ないはずです．
   ATF側にもなにか対応が必要かもしれません．
 ・doscomの状態
     ・初期化してしまうと通信が断絶するため，継続するならコントロールブロックを特別な領域に置く等が必要になるかと思います．

**Reply by Explorer-kato** [reply]
お世話になっております。
ご回答いただいた内容につきまして、まずは更新する前にある程度再起動できないか試しております。

exit_kernelの終了処理ルーチンについて質問させていただきたいのですが
以下の処理でプログラムが固まるように見受けられたので中を調べてみたところ
terrtnbb_table[0]が0x00000003という値になっているのが原因のように見受けられました。
	if (is_mprc(p_my_pcb)) {
		call_terrtn(&(terrtnbb_table[0]));
	}
直前のterrtnbb_table[p_my_pcb->prcid※1と想定]は0xbe010fd0で_kernel_terrtnb_table_gloを示しているようです。
グローバル終了処理ルーチンと思われるterrtnbb_table[0]には何が入るべきでしょうか？
本処理をコメントアウトしても処理が固まるのは変化しませんでした。

目的のいたしましてはこの処理を抜けた後の
target_exit()にToppersのエントリへのジャンプ処理を入れて動作を確認したいと考えており
imx8mm_evk.ldのPROVIDE(software_term_hook = 0);をPROVIDE(software_term_hook = 0xBE000000);とすることで
以下の処理でエントリへジャンプできるのではないかと考えております。
	if (fp != 0) {
		(*fp)();
	}
もし間違いがございましたらご指摘いただけますと幸いです。

使用パッケージは更新の質問させていただいたものと同じものを使用しております。

以上です。よろしくお願いいたします。

**Reply by hiroakitakada** [reply]
まず，terrtnbb_tableについて回答します。
このテーブルは，TERRTNBB型の配列で，TERRTNBB型は以下の定義になっています。
```
typedef struct termination_routine_block_table_block {
    uint_t          tnum_terrtn;        /* 終了処理ルーチンの数 */
    const TERRTNB   *terrtnb_table;     /* 終了処理ルーチンブロックテーブル */
} TERRTNBB;
```

>  terrtnbb_table[0]が0x00000003という値になっているのが原因のように見受けられました。

これは，terrtnbb_table[0].tnum_terrtnが3ということではないでしょうか？ もしそうであれば，グルーバル終了処理ルーチンが3つ登録されているということです。

> 直前のterrtnbb_table[p_my_pcb->prcid※1と想定]は0xbe010fd0で_kernel_terrtnb_table_gloを示しているようです。

こちらは，terrtnbb_table[0].terrtnb_tableがグルーバル終了処理ルーチンのテーブル（先頭番地と拡張情報からなる構造体の配列）の先頭番地を指しているということだと思います。

以上から，いただいた情報からは，terrtnbb_tableはおかしくないように思います。

**Reply by hiroakitakada** [reply]
> imx8mm_evk.ldのPROVIDE(software_term_hook = 0);をPROVIDE(software_term_hook = 0xBE000000);とすることで
> 以下の処理でエントリへジャンプできるのではないかと考えております。

リンカスクリプト中のPROVIDEは，このシンボルに対するweak definition（他にシンボル定義がなければ，この定義が有効になる）を与えるものです。ですので，リンカスクリプトを書き換えずに，software_term_hookというシンボルを別に定義することで，このシンボルを上書きすることができます。具体的には，software_term_hookという関数（または，アセンブリ言語ルーチン）を作り，そこから必要な番地にジャンプさせれば良いです。

**Reply by Explorer-kato** [reply]
早速のご回答ありがとうございます。
terrtnbb_tableにつきまして理解できました。
現状safeg64_1.2.0\apps\sample_sのアプリを使用してQコマンドを実行した後の終了処理で
以下のログで固まってしまいます。プロセッサ数は1としています。
-- buffered messages --
Sample program ends.
local_terrtn exinf = 1, counter = 2
global_terrtn exinf = 0, counter = 3

call_terrtn(&(terrtnbb_table[0]));の下にログを追加しているのですがそちらも表示されないため
本箇所が原因と考えており、terrtnbb_tableが怪しいと考えていました。

本来であればtarget_exit関数へ飛び、追加ログでJump Entry Point を表示させたいのですが
表示させることもできません。原因に何かお心当たりはございますでしょうか？
	if (fp != 0) {
		syslog(LOG_NOTICE, "Jump Entry Point %08X", (volatile unsigned int)fp);
		(*fp)();
	}

software_term_hookの件につきましてもありがとうございます。
終了後に再度エントリポイントにジャンプさせる場合、
chip_terminate();の後で行った方がよろしいでしょうか？

**Reply by hiroakitakada** [reply]
> 本来であればtarget_exit関数へ飛び、追加ログでJump Entry Point を表示させたいのですが
> 表示させることもできません。原因に何かお心当たりはございますでしょうか？
> if (fp != 0) {
> syslog(LOG_NOTICE, "Jump Entry Point %08X", (volatile unsigned int)fp);
> (*fp)();
> }

LOG_NOTICEをLOG_EMERGに変更してみてください。LOG_NOTICEだと，まずシステムログバッファに入ります。システムログバッファに入ったものを，グローバル終了処理ルーチンとして登録されているlogtask_terminateでシリアルI/Oに出力しています。そのため，LOG_NOTICEですと，グローバル終了処理ルーチンより後に出されたものは，出力されません。LOG_EMERGですと，syslogを呼んだ中でシリアルI/Oに出力しますので，出力される可能性が高いです。

**Reply by hiroakitakada** [reply]
> software_term_hookの件につきましてもありがとうございます。
> 終了後に再度エントリポイントにジャンプさせる場合、
> chip_terminate();の後で行った方がよろしいでしょうか？

software_term_hookでエントリポイントにジャンプさせるか，target_exitの最後でエントリポイントにジャンプさせるかですが，違いはchip_terminateを呼ぶか呼ばないかだけですので，終了処理・初期化処理をちゃんと設計すれば，どちらで実施しても良いと思います。

**Reply by Explorer-kato** [reply]
ご回答ありがとうございます。
LOG_EMERGに変更してみたのですが途中で止まっていたと報告いたしましたが
調べたところ以下が呼ばれていてそもそも出力を止めているようでした。
logtask_terminate
sio_terminate
ですのでジャンプ処理まで進んでいる想定で確認を進めたいと思います。
お騒がせしてすみません。

software_term_hookの件設計次第でどちらでもいいということで承知いたしました。
ちなみにimx8mm_evkの環境でサンプルを使用した検証するのであればchip_terminateを含めた方がよいでしょうか？

**Reply by hiroakitakada** [reply]
> LOG_EMERGに変更してみたのですが途中で止まっていました。
> call_terrtn関数にログ出力を追加したところログの内容自体は全部出せているのですが
> 関数抜けた後のログが表示されておりません。

ログが出ない原因として気になるのは，sio_terminate（シリアルドライバの終了処理）を終了処理ルーチンとして登録してあることです。これを呼び出した結果，シリアル出力が止まってしまうのかもしれません（そのターゲットでsio_terminateがどういう中身になっているか次第です）。システムコンフィギュレーションファイルから，sio_terminateのATT_TERを削除してみてください。

>  ちなみにimx8mm_evkの環境でサンプルを使用した検証するのであればchip_terminateを含めた方がよいでしょうか？

chip_terminateで何を実施しているか次第だと思いますが，imx8mmのターゲット依存部を見ていないので，私には何とも言えません。

**Reply by Explorer-kato** [reply]
ご回答ありがとうございます。
投稿内容の編集と入れ違いになってしまいましたが
sio_terminateが呼ばれているのがログが出なくなる原因と解釈いたしました。

**Reply by Explorer-kato** [reply]
度々のご連絡失礼いたします。
safeg64_1.2.0のtarget_user.txtに従って(私の環境はimx8mmです)
apps/sample_sで生成されるfmp.binを使用したSafeG64の動作確認につきまして、
Toppersで’Q’コマンド実行して停止した際、
Linuxも同時に止まってしまう挙動は問題ありますでしょうか？私の環境では止まります。
挙動に問題ない場合、原因や対策可能であれば対策についてご教示いただけないでしょうか？

現状Toppersの更新確認のために0xBE800000をエントリポイントとしたfmp.binを作成、
Linuxの自作のアプリからそのファイルを0xBE800000~に書き込み、
起動時に動作しているToppersはtarget_exit()で0xBE800000にジャンプするような処理にしております。
その状態で'Q'コマンドを実行すると各プロセッサが"0xBE800000へジャンプします”、というログを表示し
Linux、Toppersもコンソール操作不可となります。
本件が上記原因と関係していないかと思い質問させていただきました。

target_exit関数内のチップ依存の終了処理につきましては
Global Interrupt Controllerの制御を行っていたので、
0xBE800000へのジャンプ前にある本処理をコメントアウトしたのですが、
コメントアウト前と比べ現象に変化ありませんでした。

更新用のToppersにつきましては、更新前でsio_terminateのATT_TERを削除しているため
sioの初期化処理をコメントアウトしております。
更新前のチップ依存の終了処理をコメントアウトしている場合は
Global Interrupt Controllerの初期化処理をコメントアウトしております。

**Comment by Explorer-kato** [comment]
ご回答ありがとうございます。
②の方法でタスク内等の簡単な処理については変更後に挙動も切り替わるのが確認できたのですが
結局タスクやハンドラそのものが変化してしまうと対応できないことが分かったので
①の方法を検討していました。いただいた情報を元に検証させていただきます。
何かありましたらまた、相談させていただければと思います。
